### PR TITLE
Migrate to Spark 2.0

### PR DESCRIPTION
**Pull request by vidma**
This migrates `spark-notebook` to `Spark 2.0` dropping support from earlier Spark versions . 
for #708 

some testing to be done
- [x] [using deprecated HiveContext works OK](https://gist.github.com/vidma/1da51beddd6e2503420adebace3e2975)
- [x] check main widgets: PivotChart
- [x] check DownloadCsv (API changed,  consolidation into 1 file & rename broken) #724

just Ignore `merge` commit mess for now.

-----

@andypetrella so what's the plan for future? do we drop support of spark 1.6 and leave spark 2.0 only?

cc @astrauka

### CODE REVIEW

File: modules/common/src/main/scala/notebook/front/widgets/Dataset.scala [hunk 1]
```diff
@@ -37,7 +36,7 @@ trait DataFrameView {
     override implicit def singleCodec: Codec[JsValue, JsValue] = JsonCodec.idCodec
   }
 
-  private lazy val json: RDD[(String, Long)] = data.toJSON.zipWithIndex.cache
+  private lazy val json: RDD[(String, Long)] = data.toJSON.rdd.zipWithIndex.cache
```
Comment (vidma) [inline modules/common/src/main/scala/notebook/front/widgets/Dataset.scala]: @andypetrella this implicit `.cache` call shouldn't be here, right (not related to this PR) ?

File: modules/common/src/main/scala/notebook/front/widgets/Dataset.scala [hunk 2]
```diff
@@ -61,12 +60,12 @@ trait DataFrameView {
   }
 }
 
-class DataFrameWidget(
-  override val data: DataFrame,
+class DatasetWidget[T](
+  override val data: Dataset[T],
   override val pageSize: Int = 25,
   extension: String
```
Comment (vidma) [inline modules/common/src/main/scala/notebook/front/widgets/Dataset.scala]: what is tis `extension = "consoleDir"` param for?

File: project/Dependencies.scala [hunk 3]
```diff
@@ -28,116 +28,74 @@ object Dependencies {
...
-      |*                                                                                                 *
-      |***************************************************************************************************
-      """.stripMargin)
-    sys.exit(1)
-  }
+  val defaultSparkVersion = sys.props.getOrElse("spark.version", "2.0.0")
 
   val sparkVersionTuple = defaultSparkVersion match { case extractVs(v, m, p) =>  (v.toInt, m.toInt, p.toInt)}
-  val defaultScalaVersion = sys.props.getOrElse("scala.version", "2.10.5") match {
+  val defaultScalaVersion = sys.props.getOrElse("scala.version", "2.10.6") match {
```
Comment (vidma) [inline project/Dependencies.scala]: with Spark 2.0 maybe we should default to Scala 2.11 ?

**Comment by vidma** [comment]
this is basically good enough now.
a bit more testing and we can 🚢  it.

@andypetrella @astrauka

**Comment by astrauka** [comment]
:clap:

**Comment by vidma** [comment]
Forgot to merge the commits, so squashed in #727  (looks like github is amazingly good at squashing that horrible merge commit noise)

File: modules/common/src/main/scala/notebook/front/widgets/Dataset.scala (hunk 1 above)
Comment (andypetrella) [inline modules/common/src/main/scala/notebook/front/widgets/Dataset.scala]: fyi, this was for the paging to be a bit optimized, but fine to remove it

File: conf/scripts/init.sc [hunk 4]
```diff
@@ -10,12 +10,15 @@ import notebook.front.widgets.magic._
...
-
 import org.apache.spark.{SparkContext, SparkConf}
 import org.apache.spark.SparkContext._
 import org.apache.spark.rdd._
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions._
 
 import scala.util.matching.Regex
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
```
Comment (andypetrella) [inline conf/scripts/init.sc]: 👍

**Pull request merged**